Ensemble-Model.md

```diff
@@ -0,0 +1,1066 @@
+        # Addition of 0.001 to the continuous features avoids a zero value from the
+        # categorical variables, allowing categorical features precedence.
+        distances = (cont_features + 0.001) * cat_features
+
+        # Compute the LR based on precomputed distributions that relate the score
+        # to a probability of providing good addon recommendations.
+        lrs_from_scores =\
+            np.array([self.get_lr(distances[i]) for i in range(self.num_donors)])
+
+        # Sort the LR values (descending) and return the sorted values together with
+        # the original indices.
+        indices = (-lrs_from_scores).argsort()
+        return lrs_from_scores[indices], indices
+
+    def get_weighted_recommendations(self, client_data):
+        recommendations = defaultdict(int)
+
+        for donor_score, donor in zip(*self.get_similar_donors(client_data)):
+            for addon in self.donors_pool[donor]['active_addons']:
+                recommendations[addon] += donor_score
+        
+        return recommendations
+```
+
+### Locale Recommender
+
+Depends on number of installs in that locale
+
+
+```python
+from taar.recommenders import LocaleRecommender
+```
+
+
+```python
+TOP_ADDONS_BY_LOCALE_FILE_PATH = "top_addons_by_locale.json"
+```
+
+
+```python
+class NewLocaleRecommender(LocaleRecommender):
+    def __init__(self, TOP_ADDONS_BY_LOCALE_FILE_PATH):
+        OriginalLocaleRecommender.__init__(self)
+        
+        with open(TOP_ADDONS_BY_LOCALE_FILE_PATH) as data_file:
+            top_addons_by_locale = json.load(data_file)
+            
+        self.top_addons_by_locale = defaultdict(lambda: defaultdict(int), top_addons_by_locale)
+        
+    def get_weighted_recommendations(self, client_data):
+        client_locale = client_data.get('locale', None)
+        return defaultdict(int, self.top_addons_by_locale[client_locale])
+```
+
+### Legacy Recommender
+
+1 for all replacement addons, 0 otherwise
+
+
+```python
+from taar.recommenders import LegacyRecommender
+```
+
+
+```python
+class NewLegacyRecommender(LegacyRecommender):
+    def get_weighted_recommendations(self, client_data):
+        recommendations = defaultdict(int)
+        addons = client_data.get('disabled_addons_ids', [])
+        
+        for addon in addons:
+            for replacement in self.legacy_replacements.get(addon, []):
+                recommendations[replacement] += 1
+                
+        return recommendations
+```
+
+## Choosing training, validation and test sets
+
+For training and validation purposes, only clients that have WebExtension addons installed are useful.
+
+
+```python
+useful_clients = completed_rdd.filter(lambda client: len(client['installed_addons']) >= 1).cache()
+```
+
+
+```python
+useful_clients.count()
+```
+
+
+
+
+    435313
+
+
+
+These users are useful for training and evaluating our model:
+
+
+```python
+training, validation, test = useful_clients.randomSplit([0.8, 0.1, 0.1])
```

Comment:
perhaps try something closer to [0.6, 0.2, 0.2] for a larger number of points evaluated against the cost function.

Reply:
I'm not using the validation set anyways right now, so I changed it to `[0.8, 0, 0.2]` for now. 

Originally, I was planing to use a validation set for things like early stopping. I didn't really have enough time for that though and it doesn't seem like overfitting by training for too long is something that happens here anyways.